XiEditor/FileSystemItem.swift

```diff
@@ -0,0 +1,119 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Cocoa
+
+final class FileSystemItem {
+
+    fileprivate var relativePath: String!
+    fileprivate var parent: FileSystemItem?
+    lazy fileprivate var children: [FileSystemItem] = self.getChildren()
+
+    var fullPath: String {
+        guard
+            let parent = self.parent,
+            let url = URL(string: parent.fullPath)?.appendingPathComponent(self.relativePath)
+            else { return self.relativePath }
+
+        return url.absoluteString
+    }
+
+    var url: URL {
+        return URL(string: self.fullPath)!
+    }
+
+    var fileUrl: URL {
+        return URL(string: "file://" + self.fullPath)!
+    }
+
+    var fileType: String {
+        return self.fileUrl.pathExtension
+    }
+
+    var isDirectory: Bool {
+        var isDir: ObjCBool = false
+        FileManager.default.fileExists(atPath: self.fullPath, isDirectory: &isDir)
+        return isDir.boolValue
+    }
+
+    convenience init() {
+        self.init(path: "/", parent: nil)
+    }
+
+    init(path: String, parent: FileSystemItem?) {
+        self.relativePath = URL(fileURLWithPath: path).lastPathComponent
+        self.parent = parent
+    }
+
+    // MARK: - Public methods
+
+    public func getChild(at index: Int) -> FileSystemItem {
+        return children[index]
+    }
+
+    public func getNumberOfChildren() -> Int {
+        return children.count
+    }
+
+    public func getName() -> String {
+        return self.relativePath
+    }
+
+    // MARK: - Private methods
+
+    private func getChildren() -> [FileSystemItem] {
+        let fileManager = FileManager.default
+        var isDirectory: ObjCBool = false
+        var children = [FileSystemItem]()
+
+        let valid = fileManager.fileExists(atPath: self.fullPath, isDirectory: &isDirectory)
+
+        if valid && isDirectory.boolValue {
+            if let contents = try? fileManager.contentsOfDirectory(atPath: self.fullPath) {
+                contents.forEach {
+                    // Don't add .DS_Store files. Probably should do a more rigorous
+                    // check here, and only allow certain types of files
+                    if ![".DS_Store"].contains($0) {
+                        children.append(FileSystemItem(path: $0, parent: self))
+                    }
+                }
+            }
+            return children
+        }
+        return []
+    }
+
+    // MARK: - Static methods
+
+    static func createParents(url: URL) -> FileSystemItem {
+        let parentUrl = url.deletingLastPathComponent()
+        let path = parentUrl.absoluteString
```

Comment:
Possibly a redundant observation at this point, but Swift API design guidelines dictates that things like URL (which is an initialism) should be uniformly be upper or lowercased. So right here, `url` can be lowercase but `parentUrl` should be changed to `parentURL`.